session.go

```diff
@@ -814,9 +845,11 @@ func (q *Query) execute(conn *Conn) *Iter {
 }
 
 func (q *Query) attempt(keyspace string, end, start time.Time, iter *Iter, host *HostInfo) {
-	q.attempts++
-	q.totalLatency += end.Sub(start).Nanoseconds()
-	// TODO: track latencies per host and things as well instead of just total
+	hostMetrics := q.getHostMetrics(host)
+	q.session.mu.Lock()
+	hostMetrics.attempts++
+	hostMetrics.totalLatency += end.Sub(start).Nanoseconds()
```

Comment:
I'd keep track of the total number of attempts and total latency in here (as per my recommendation to keep the fields on `Query` above) and also calculate the average as there's already an attempt, thereby you can do away with the check for 0 division. That way, you can get rid of the lock dance in both `Latency()` and `Attempts()` as well as the for loops and so on.

Reply:
That'll simplify the code above and because you're protecting the recording of the attempt with the lock anyway it's easy track of these values without fear they'll drift.